### DISCUSSION TITLE
In-chapter page numbers

### DISCUSSION THREAD

**Discussion by emagugiatti** [QuestPDF/QuestPDF#120] [Ideas]
Having the need for in-chapter page numbering, I tried taking existing classes and functions and turning them in what I needed, which resulted in something like the following.

In the ElementExtensions class I added a new method called 'ChapterStart', which leverages the possibility to add a location to the document, but automatically prefixing "Ch_" to have a clear distinction between these and 'regular' locations.
```
public static IContainer ChapterStart(this IContainer element, string chapterName)
{
  return element.Element(new InternalLocation
  {
    LocationName = "Ch_" + chapterName
  });
}
```

Then, in the TextDescriptor class I took 'CurrentPageNumber' and turned it into 'PageNumberInChapter':
```
public void PageNumberInChapter(string chapterName, TextStyle? style = null)
{
  if (IsNullOrEmpty(chapterName))
    throw new ArgumentException(nameof(chapterName));

  ChapterPageNumber(chapterName, style);
}
```
which calls a modified version of 'PageNumber' called 'ChapterPageNumber'
```
private void ChapterPageNumber(string slotName, TextStyle? style = null)
{
  if (IsNullOrEmpty(slotName))
    throw new ArgumentException(nameof(slotName));

  style ??= TextStyle.Default;

  AddItemToLastTextBlock(new TextBlockChapterPageNumber()
  {
    Style = style,
    SlotName = slotName
  });
}
```
that references to 'TextBlockChapterPageNumber', a modified 'TextBlockPageNumber' (_who would have guessed_), basically identical apart from the SetPageNumber function:
```
private void SetPageNumber(IPageContext context)
{
  var pageNumberPlaceholder = 999;
  var startPageNumber = context.GetRegisteredLocations().Contains("Ch_" + SlotName) //page number where chapter starts
      ? context.GetLocationPage(SlotName)
      : pageNumberPlaceholder;

  var currentPageNumber = context.GetRegisteredLocations().Contains(PageContext.CurrentPageSlot)
      ? context.GetLocationPage(PageContext.CurrentPageSlot)
      : pageNumberPlaceholder;

  var currentChapterPage = currentPageNumber - startPageNumber + 1; //calculating how many pages between current and start of chapter

  Text = currentChapterPage.ToString();
}
```

In the end, the user needs to set a ChapterStart location where a chapter starts in the document with `text.ChapterStart("Chapter5")` and then `text.PageNumberInChapter("Chapter5")` to get the page number relative to the start of chapter they set.

I know, it's far from perfect, should probably use new elements instead of just reusing renamed InternalLocations, but I did it quickly and just wanted to give you my suggestion for something that might be helpful.
Also, the possibility to set an end to a chapter and calculate and print the total pages of it would be cool (something like 'Chapter 5: page 5 of 16' or something like that).

Thanks!

**Comment by MarcinZiabek** [comment]
An alternative solution would be to extend/modify the page concept of the document.

```csharp
public class StandardReport : IDocument
{
    // metadata

    public void Compose(IDocumentContainer container)
    {
        container
            .Page(page =>
            {
                page.SectionName("chapter1"); // <-- and here
                page.MarginVertical(80);
                page.MarginHorizontal(100);
                page.Background(Colors.Grey.Medium); // transparent is default
                page.Size(PageSizes.A3);
                    
                page.Header().Element(ComposeHeader);
                page.Content().Element(ComposeBigContent);
                page.Footer().AlignCenter().PageNumber();
            })
            .Page(page =>
            {
                // you can specify multiple page types in the document
                // with independend configurations
                page.SectionName("chapter2"); // <-- and here
                page.Margin(50)
                page.Size(PageSizes.A4);
                    
                page.Header().Element(ComposeHeader);
                page.Content().Element(ComposeSmallContent);
                page.Footer().AlignCenter().PageNumber();
            });
    }

    // content implementation
}
```

And call it like this:
```csharp
text.CurrentPageNumberWithinSection("chapter1");
text.TotalPagesOfSection("chapter2");
```

This way, we can clearly divide entire document into sections. The section content describes its beginning and end. The API looks promising but have a couple of limitations:
1) Sections/chapters cannot intersects - does it have sense to have intersections though?
2) Sections need to be implemented on separate groups of pages. A single page cannot be occupied by multiple sections.

**Reply by emagugiatti** [reply]
Yup, that makes way more sense, also fully agree with your two points, doesn't make sense to have multiple sections on a single page since a new one would start with a new page, and an instersection, even if accidental, is something we don't want.

Thanks for your help, looking forward for the future updates!

**Comment by MarcinZiabek** [comment]
I was thinking about this requirement and I am considering slightly different option:

1. Add following methods to the Text API:

```csharp
.BeginPageNumberOfLocation("name");
.EndPageNumberOfLocation("name");
.PageNumberWithinLocation("name");
.PageLengthOfLocation("name");
```

2. Deprecate the following method:

```csharp
// we now take into account begin AND end of the location
.PageNumberOfLocation("name");
```

3. Extend all page-ralated methods by an optional formatting argument.

```csharp
// Func<int?, string> that maps page number to the its representation
// Nullable<int> because in the first rendering pass, we don't know exact page numbers, best to put a placeholder
// This approach also solves strange requirements where page numbering should start at value different than 1, just add proper mapping whenever calling page-related API
.PageLengthOfLocation("name", x => x == 1 ? "1 page long" : $"{x} pages long");
```

It is good to think about argument order or using overload:

```csharp
void PageLengthOfLocation(string locationName, TextStyle? style = null);
void PageLengthOfLocation(string locationName, Func<int?, string>? format = null, TextStyle style = null);
void PageLengthOfLocation(string locationName, TextStyle? style = null, Func<int?, string>? format = null); // best one in my opinion as I expect that custom format is very rarely used feature
```

4. To describe locations, use already existing `InternalLocation` API method.

5. Maybe we should:
    - raname `InternalLink` to `SectionLink`,
    - rename `InternalLocation` to `Section`. This also involves renaming of new methods, e.g. to `BeginPageNumberOfSection`, `PageNumberWithinSection`, etc.

---

Example usage:

```csharp
.Text(text =>
{
    text.BeginPageNumberOfLocation(locationName);
    text.Span(" - ");
    text.EndPageNumberOfLocation(locationName);

    var lengthStyle = TextStyle.Default.Color(Colors.Grey.Medium);
    
    text.Span(" (", lengthStyle);
    text.PageLengthOfLocation(locationName, x => x == 1 ? "1 page long" : $"{x} pages long", lengthStyle);
    text.Span(")", lengthStyle);
});
```

Result:
![image](https://user-images.githubusercontent.com/9263853/153959075-c317c68e-d232-4b5e-bfe9-48e0f3c6106c.png)

---

What do you think? The goal is not only to add new features but also slowly refine the API so it is easier to discover. Sometimes however, it is challenging to find best direction of changes.

@wesleigomes @marcianobandeira @qcz @lmingle @jcl-aadlab @donmurta @adaxer @pha3z  I am allowing myself to include you to this discussion, thank you in advance for any suggestions.

**Comment by emagugiatti** [comment]
That's exactly what I was thinking about! And the usage it's intuitive enough, both for new users of this library and old ones that updating to a new version.

Just a question about points 2 and 5 though: by turning an `InternalLocation` into a `Section` it would be necessary to define both 'Begin' and 'End' points, but what if we do not _need_ to define beginning and end and just set a link to a point in the document? Would it make sense to keep `InternalLocations` for just `InternalLinks`? Maybe for something like footnotes, referencing a particular word, phrase or quote in the document content, or even the reverse of this, like Wikipedia does with its citations/notes, that, when clicked bring you directly to the bottom of the page/document where you can find extensive links to the sources.

Fully agree with your other points. Thanks.

**Reply by MarcinZiabek** [reply]
In my idea, the `Section` area is defined by its child (text, image, column, table, anything). If it contains a lot of content, the section may span multiple pages. Therefore, there is no need to define beginning and end, as the content itself acts as a definition. The `SectionLink` will always redirect to the beginning of the section (page + starting point). And SectionLink behaves very similar. It also can have various content as a child (column, image, text), and anything within that content is clickable.

```csharp
.Column(column => 
{
    // possibly many pages long
    column.Item().Section("mySection").Grid(grid => 
    {
         foreach(var i of Enumerable.Range(1, 100)
             grid.Item().Text($"This is line number {i}");
    });

    // this image redirects to the beginning of mySection
    column.Item().SectionLink("mySection").Image(Placeholders.Image(100, 200));
});
```

So, in terms of functionality, `Section` is an extension of `Location`. It will provide exactly the same set of functionality AND add support for:
- tracking not only section start page but also its entire content, with end page,
- give access to page number withing section (like section resets page numbering),
- give access to length of the section (how many pages it spans).

The purpose or renaming, in my opinion, is simple: `Location` suggests that you are focused on a single place/point whereas `Section` gives you a hint that it can define a bigger part of the document if needed. I am not sure if there is a point of keeping Location because underneath it will just call Section.

**Reply by emagugiatti** [reply]
Oh I see it, I misunderstood before. Then yes, `Location` now would be redundant given the more possibilities of `Section`. Awesome, thank you.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
